src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -10610,6 +10592,84 @@ private bool IsTargetTypedObjectCreation()
             }
         }
 
+#nullable enable
+
+        private ExpressionSyntax ParseWithExpression(ExpressionSyntax receiverExpression, SyntaxToken withKeyword)
+        {
+            var openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
+
+            var assignments = _pool.AllocateSeparated<AssignmentExpressionSyntax>();
+            if (CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+            {
+tryAgain:
+                if (CurrentToken.Kind == SyntaxKind.IdentifierToken)
+                {
+                    // First assignment
+                    assignments.Add(parseAssignment());
+
+                    // Rest
+                    while (CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+                    {
+                        if (CurrentToken.Kind == SyntaxKind.CommaToken || IsNamedAssignment())
+                        {
+                            assignments.AddSeparator(EatToken(SyntaxKind.CommaToken));
+
+                            if (CurrentToken.Kind == SyntaxKind.CloseBraceToken)
+                            {
+                                break;
+                            }
+                            assignments.Add(parseAssignment());
+                        }
+                        else if (skipBadTokens(ref openBrace, SyntaxKind.CommaToken) == PostSkipAction.Abort)
+                        {
+                            break;
+                        }
+                    }
+                }
+                else if (skipBadTokens(ref openBrace, SyntaxKind.IdentifierToken) == PostSkipAction.Continue)
+                {
+                    goto tryAgain;
+                }
+            }
+            var closeBrace = this.EatToken(SyntaxKind.CloseBraceToken);
+
+            withKeyword = CheckFeatureAvailability(withKeyword, MessageID.IDS_FeatureRecords);
+
+            var result = _syntaxFactory.WithExpression(
+                receiverExpression,
+                withKeyword,
+                openBrace,
+                assignments.ToList(),
+                closeBrace);
+
+            _pool.Free(assignments);
+
+            return result;
+
+            AssignmentExpressionSyntax parseAssignment()
+            {
+                var identifier = ParseIdentifierName();
+                var equal = EatToken(SyntaxKind.EqualsToken);
+                var expression = ParseExpressionCore();
+
+                return _syntaxFactory.AssignmentExpression(
+                    SyntaxKind.SimpleAssignmentExpression,
+                    identifier,
+                    equal,
+                    expression);
+            }
+
+            PostSkipAction skipBadTokens(ref SyntaxToken startToken, SyntaxKind expected)
+            {
+                return this.SkipBadSeparatedListTokensWithExpectedKind(ref startToken, assignments,
+                    p => p.CurrentToken.Kind != SyntaxKind.IdentifierToken && !p.IsNamedAssignment(),
```

Comment:
>IsNamedAssignment [](start = 81, length = 17)

If the token kind is not identifier, then IsNamedAssignment always returns false. In other words, I think the second half of this `&&` operation does not contribute the the result.